MatrixSDK/Crypto/Verification/MXDeviceVericationManager.m

```diff
@@ -0,0 +1,468 @@
+{
+    [self cancelTransaction:transaction.transactionId fromUserId:transaction.otherUserId andDevice:transaction.otherDeviceId code:code];
+}
+
+- (void)cancelTransaction:(NSString*)transactionId fromUserId:(NSString*)userId andDevice:(NSString*)deviceId code:(MXTransactionCancelCode*)code
+{
+    MXKeyVerificationCancel *cancel = [MXKeyVerificationCancel new];
+    cancel.transactionId = transactionId;
+    cancel.code = code.value;
+    cancel.reason = code.humanReadable;
+
+    NSLog(@"[MXKeyVerification] cancelTransaction: transactionId: %@. Code:%@. Reason: %@", transactionId, cancel.code, cancel.reason);
+
+    [self sendToOther:userId deviceId:deviceId eventType:kMXEventTypeStringKeyVerificationCancel content:cancel.JSONDictionary success:nil failure:^(NSError *error) {
+
+        NSLog(@"[MXKeyVerification] cancelTransaction. Error: %@", error);
+    }];
+
+    [self removeTransactionWithTransactionId:transactionId];
+}
+
+
+#pragma mark - Incoming to_device events
+
+/**
+ Handle a to-device event.
+
+ @param notification the notification containing the to-device event.
+ */
+- (void)onToDeviceEvent:(NSNotification *)notification
+{
+    MXEvent *event = notification.userInfo[kMXSessionNotificationEventKey];
+    
+    dispatch_async(cryptoQueue, ^{
+        switch (event.eventType)
+        {
+            case MXEventTypeKeyVerificationStart:
+                [self handleStartEvent:event];
+                break;
+
+            case MXEventTypeKeyVerificationCancel:
+                [self handleCancelEvent:event];
+                break;
+
+            case MXEventTypeKeyVerificationAccept:
+                [self handleAcceptEvent:event];
+                break;
+
+            case MXEventTypeKeyVerificationKey:
+                [self handleKeyEvent:event];
+                break;
+
+            case MXEventTypeKeyVerificationMac:
+                [self handleMacEvent:event];
+                break;
+
+            default:
+                break;
+        }
+    });
+}
+
+- (void)handleStartEvent:(MXEvent*)event
+{
+    NSLog(@"[MXKeyVerification] handleStartEvent");
+
+    MXKeyVerificationStart *keyVerificationStart;
+    MXJSONModelSetMXJSONModel(keyVerificationStart, MXKeyVerificationStart, event.content);
+
+    if (!keyVerificationStart)
+    {
+        return;
+    }
+    if (!keyVerificationStart.isValid)
+    {
+        if (keyVerificationStart.transactionId && keyVerificationStart.fromDevice)
+        {
+            [self cancelTransaction:keyVerificationStart.transactionId
+                         fromUserId:event.sender
+                          andDevice:keyVerificationStart.fromDevice
+                               code:MXTransactionCancelCode.invalidMessage];
+        }
+
+        return;
+    }
+
+
+    MXDeviceVerificationTransaction *existingTransaction = [self transactionWithUser:keyVerificationStart.transactionId andDevice:keyVerificationStart.fromDevice];
```

Comment:
Param is withUser, but value passed is a transaction id

Reply:
Also this test is also done after the key download, maybe this one is not needed (unless for saving a download key)?